# Analysis of specific junction

We are exemplary considering junction CIRMBUYJFK_f__CWCCKOQCWZ_r

In [1]:
# set working directory to project folder
import sys, os, re
sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))

import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.graph_objects as go

import mplcursors

import altair as alt
from itertools import combinations
import numpy as np
import pypangraph as pp
from Bio import Phylo, SeqIO, AlignIO
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord

import plotly.io as pio
pio.renderers.default = "browser" 


from pathlib import Path
import subprocess

from Bio.Phylo.TreeConstruction import DistanceCalculator
from scipy.cluster.hierarchy import linkage, dendrogram, fcluster
from scipy.spatial.distance import squareform

from junction_analysis.helpers import get_tree_order, convert_gbk_fasta, convert_gbk_fasta_plasmids
import junction_analysis.pangraph_utils as pu
from junction_analysis.plotting import plot_junction_pangraph_combined, plot_junction_pangraph_interactive, plot_dendrogram
from junction_analysis.consensus import find_consensus_paths, make_deduplicated_paths
from junction_analysis.block_alignment import create_block_msas, summarize_block_msas, analyze_alignment, cluster_alignment, retrieve_cluster_assignments

from junction_analysis.annotate_insertions import write_insertions_fasta, get_insertions_deletions_from_consensus, retrieve_SAMids_txt, combine_NCBI_atb_results, find_insertion_hits_own_genome, find_insertion_hits_in_plasmids

In [2]:
example_junction = "CIRMBUYJFK_f__CWCCKOQCWZ_r"
example_pangraph = pp.Pangraph.from_json(f"../results/junction_pangraphs/{example_junction}.json")

aln_path = f"../results/block_alignments/{example_junction}/block_11400043001338627984_aln.fa"
distance_matrix, Z, names = cluster_alignment(aln_path)
clusters = retrieve_cluster_assignments(Z, names, n_clusters=2)

## Analyze deviations (insertions / deletions) from consensus path
Consider the two consensus paths seperately from each other.

In [3]:
# choosing higher thresholds merges the two consensus paths into one, however this might not be possible for all junctions while still keeping all desired consensus paths
consensus_paths, path_dict, deduplicated_blog_freq, similarity_matrix, assignment_df = find_consensus_paths(example_pangraph, rare_block_threshold=10, rare_edge_threshold=10, min_n_isolates_per_consensus=5)
plot_junction_pangraph_interactive(
    example_pangraph,
    show_consensus=True,
    consensus_paths=consensus_paths,
    assignments=assignment_df,
    order="tree",
    cluster_map=clusters,
    title = "Junction Block Structure with 2 Clusters"
)

Found 2 unique paths.


### Consensus 2

In [4]:
insertions_2, deletions_2 = get_insertions_deletions_from_consensus(example_pangraph, assignment_df, consensus_paths, consensus = 2)
write_insertions_fasta(example_junction, example_pangraph, insertions_2, consensus = 2)

# upload fastas: scp -r /Users/mariebecker/Documents/Uni/ETH/Masterarbeit/marie-junctions/results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus2 becker0010@transfer12.scicore.unibas.ch:/scicore/home/neher/GROUP/data/2025_all_the_bacteria/lexicmap_index/queries/CIRMBUYJFK_f__CWCCKOQCWZ_r/
# run:  bash run_all_lexicmap.sh ./queries/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus2 on cluster
# download results:  scp -r becker0010@transfer12.scicore.unibas.ch:'/scicore/home/neher/GROUP/data/2025_all_the_bacteria/lexicmap_index/queries/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus2/*.lexicmap.tsv' /Users/mariebecker/Documents/Uni/ETH/Masterarbeit/marie-junctions/results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus2/

Insertions:
NZ_AP022171.1 INSERTED: [18228278273347143766|+|]
NZ_CP098219.1 INSERTED: [2278011600046483881|+|]

Deletions:


In [26]:
example_pangraph.blocks[insertions_2['NZ_AP022171.1'][0].nodes[0].id].alignment.generate_alignment()[str(14783663026967800871)]

'TTTAACCTGAAAACGACTTCACTCTACGTTACCCACGATCAGGTTGAGGCGATGACGCTCGCCCAGCGAGTAATGGTGATGAACGGCGGCGTTGCCGAACAGATTGGCACGCCAGTTGAAGTGTACGAAAAGCCCGCCAGCCTGTTTGTGGCGAGTTTTATCGGCAGCCCGGCGATGAACCTACTGGCTGGCCGTGTGAATAACGAAGGTACGCATTTCGAACTGGATGGCGGTATTGCGCTGCCGCTAAACGGTGGCTACCGTCAGTATGTCGGGCGTAAAATGACCCTCGGCATTCGCCCGGAACATATAGCGCTAAGCTCGCAGGCAGAAGGTGGTGTGCCGCTGGTGATGGACACGCTGGAGATCCTCGGCGCAGATAACCTGGCGCACGGACATTGGGGCGAACAGAAGCTGGTGGTGCGACTGGCACATCAGGAGCGCCCGACGGCAGGCAGCACGCTGTGGCTGCATCTGCCGGAAAATCAGCTACATCTTTTTGATGGTGAAACAGGACAACGAGTATGAGTAACTGGCCTTATCCCCGCATCGTCGCCCATCGTGGCGGCGGTAAACTGGCCCCGGAGAACACCCTGGCAGCAATCGACGTCGGGGCAAAATACGGTCATAAGATGATCGAATTTGACGCGAAGTTATCGAAAGATGGCGAGATCTTCCTGCTCCATGACGACAATCTTGAACGCACCAGCAACGGCTGGGGCGTCGCGGGTGAACTGAACTGGCAGGATTTACTGCGCGTGGATGCAGGCAGTTGGTACAGCAAAGCGTTTAAAGGTGAGCCACTGCCGTTGCTTTCGCA'

In [ ]:
# retrieve SAM... ids for NCBI lookup
retrieve_SAMids_txt(parent_dir="../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus2")

# to lookup in NCBI run from results/atb_lookup: bash fetch_biosamples_consensus.sh CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus1

In [22]:
combine_NCBI_atb_results(parent_dir="../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus2")

In [ ]:
# not annotated as insertion sequence but this is gene content annotation:
#NZ_AP022171.1	GenBank	gene	10173	10694	.	+	.	ID=H7R27_RS10275;Name=ugpC;gene=ugpC;is_partial=False;locus_tag=H7R27_RS10275;pseudo=
#NZ_AP022171.1	GenBank	CDS	10173	10694	.	+	.	ID=H7R27_RS10275;Name=ugpC;codon_start=1;gene=ugpC;inference=COORDINATES: similar to AA sequence:RefSeq:NP_312323.2;is_partial=False;locus_tag=H7R27_RS10275;note=incomplete%3B partial in the middle of a contig%3B missing N-terminus%3B Derived by automated computational analysis using gene prediction method: Protein Homology.;product=sn-glycerol 3-phosphate ABC transporter ATP binding protein UgpC;pseudo=;transl_table=11
#NZ_AP022171.1	GenBank	gene	10691	11008	.	+	.	ID=H7R27_RS10280;Name=H7R27_RS10280;is_partial=False;locus_tag=H7R27_RS10280;old_locus_tag=WP5S18E09_20050;pseudo=
#NZ_AP022171.1	GenBank	CDS	10691	11008	.	+	.	ID=H7R27_RS10280;Name=H7R27_RS10280;go_function=GO:0008081 - phosphoric diester hydrolase activity [Evidence IEA];go_process=GO:0006629 - lipid metabolic process [Evidence IEA];codon_start=1;inference=COORDINATES: similar to AA sequence:RefSeq:NP_709219.1;is_partial=False;locus_tag=H7R27_RS10280;note=incomplete%3B partial in the middle of a contig%3B missing C-terminus%3B Derived by automated computational analysis using gene prediction method: Protein Homology.;old_locus_tag=WP5S18E09_20050;product=glycerophosphodiester phosphodiesterase family protein;pseudo=;transl_table=11
atb_hits_info_df = pd.read_csv("../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus2/NZ_AP022171.1_segment_0.hits_info.tsv", sep = "\t")
atb_hits_info_df.value_counts("organism")

organism
Escherichia coli                  187
Escherichia coli O25b:H4-ST131      4
Acinetobacter baumannii             1
Campylobacter jejuni                1
Escherichia coli KTE216             1
Escherichia coli O13/O135:H4        1
Shigella sonnei                     1
Name: count, dtype: int64

In [ ]:
# long insertion: prophage
atb_hits_info_df = pd.read_csv("../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus2/NZ_CP098219.1_segment_0.hits_info.tsv", sep = "\t")
atb_hits_info_df.value_counts("organism")

organism
Escherichia coli                       83
Escherichia coli O145:H28               3
Salmonella enterica                     2
Salmonella enterica subsp. enterica     1
Name: count, dtype: int64

### Consensus 1

In [ ]:
# get a list of all blocks that deviate from consensus paths (consider insertions and deletions!)

insertions, deletions = get_insertions_deletions_from_consensus(example_pangraph, assignment_df, consensus_paths, consensus = 1)
write_insertions_fasta(example_junction, example_pangraph, insertions, consensus = 1)

# TODO: now correct fasta files, need to rerun downstream code with correct fasta files

Insertions:
NZ_OX030701.1 INSERTED: [6932368721519041699|-|1314867519434933921_1]
NZ_CP021935.1 INSERTED: [517157241956952768|+|]
NZ_CP076693.1 INSERTED: [964531385716605116|+|]_[16982887268023976258|+|]
NZ_CP128947.1 INSERTED: [964531385716605116|+|]_[16982887268023976258|+|]
NZ_CP107114.1 INSERTED: [517157241956952768|+|]
NZ_CP107172.1 INSERTED: [517157241956952768|+|]
NZ_CP030337.1 INSERTED: [6932368721519041699|+|16056383497489254679_1]
NZ_CP124429.1 INSERTED: [517157241956952768|-|]
NZ_CP018970.1 INSERTED: [6932368721519041699|-|17115477248436968411_1]
NZ_CP019015.1 INSERTED: [16982887268023976258|-|]_[964531385716605116|-|]
NZ_CP124460.1 INSERTED: [964531385716605116|+|]_[16982887268023976258|+|]
NZ_AP022044.1 INSERTED: [6932368721519041699|+|12883103544713419939_1]
NZ_AP022044.1 INSERTED: [6932368721519041699|-|14506298656834147691_1]
NZ_CP103562.1 INSERTED: [7844090243072536214|+|]
NZ_CP051661.1 INSERTED: [6932368721519041699|+|12883103544713419939_1]
NZ_CP051661.1 INSERTED: [6

run lexicmap search on cluster for all sequences: bash run_all_lexicmap.sh ./queries/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus1

In [ ]:
# retrieve SAM... to look up in NCBI:
retrieve_SAMids_txt(parent_dir="../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus1")

retrieve info from NCBI: bash fetch_biosamples_consensus1.sh out of this folder results/atb_lookup

In [ ]:
# merge results
combine_NCBI_atb_results(parent_dir="../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus1")

In [141]:
pd.read_csv("../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus1/NZ_CP030337.1_segment_0.ncbi_results.tsv", sep="\t").value_counts('organism')

organism
Campylobacter jejuni         9
Escherichia coli             2
Helicobacter winghamensis    1
Klebsiella pneumoniae        1
Name: count, dtype: int64

In [ ]:
# TODO: should I run blocks seperately or insertions as a whole
# TODO: rerun lexicmap ... now with correctly inverted sequences

Information about different insertion sequences:

- 5 different versions of this block, if a block is twice in one sequence it is always the same version twice, versions can be inserted or not inserted, found between 1 and 9 times in other genomes
    - NZ_CP107151.1, blocks [6932368721519041699|-|]: majority of hits are Mycoplasma mycoides or E. coli and some others, IS1 based on annotation, same insertion in NZ_CP107162.1 and NZ_CP107122.1 blocks [6932368721519041699|-|], same hits, NZ_CP107122.1 even has same insertion a second time[6932368721519041699|+|], NZ_CP128907.1|segment_0 path[6932368721519041699|+|1708810210093552769] length760, NZ_AP022326.1|segment_0 path[6932368721519041699|+|5100696243500885373] length777, NZ_AP022326.1|segment_1 path[6932368721519041699|+|6097212296783908714] length777, NZ_CP124410.1: path[6932368721519041699|+|] IS1, Mycoplasma mycoides and E.coli and Campylobacter jejuni and some more
    - NZ_CP021454.1|segment_0 path[6932368721519041699|-|913783977968963568] length777 and NZ_CP019012.1|segment_0 path[6932368721519041699|-|10145330714100251892] length768 (only found once in genome), inverted, IS1, only 4 hits and different ones --> difference within block!
    - NZ_OX030701.1|segment_0 path[6932368721519041699|-|] length760, NZ_CP018970.1|segment_0 path[6932368721519041699|-|17115477248436968411] length767, NZ_CP023826.1|segment_0 path[6932368721519041699|-|17115477248436968411] length767, NZ_LR890693.1|segment_0 path[6932368721519041699|-|11588582069103229002] length760, NZ_CP051661.1|segment_1 path[6932368721519041699|+|10145330714100251892] length776 and segment_0 path[6932368721519041699|+|12883103544713419939] length776, NZ_CP051663.1|segment_0 path[6932368721519041699|+|7987278316461198430] length764, NZ_CP051659.1|segment_0 path[6932368721519041699|+|10145330714100251892] length760, NZ_CP018970.1|segment_0 path[6932368721519041699|-|17115477248436968411] length767, NZ_CP023826.1|segment_0 path[6932368721519041699|-|17115477248436968411] length767, NZ_CP124455.1|segment_0 path[6932368721519041699|-|15558722379768673797] length768, same hits, less hits, IS1
    - NZ_CP030337.1|segment_0 path[6932368721519041699|+|16056383497489254679] length760, Campylobacter jejuni version (?)
    - NZ_AP022044.1|segment_0 path[6932368721519041699|+|12883103544713419939] length767 and NZ_AP022044.1|segment_1 path[6932368721519041699|-|14506298656834147691] length767, new version, mostly E.coli, 25 hits

- NZ_CP103562.1: path[7844090243072536214|+|] E. coli and some single different ones, IS3, found twice in genome

- double block: all found twice (once elsewhere in genome), all at same position, just some inverted
    - NZ_CP124374.1 and NZ_CP019015.1: [16982887268023976258|-|]_[964531385716605116|-|] ISL3 (both blogs together): no hits (TODO: change parameters and rerun or run blogs seperately)
    - NZ_CP059130.1 (segment 1), NZ_AP026788.1 and NZ_AP026794.1, [964531385716605116|+|]_[16982887268023976258|+|], NZ_CP124460.1|segment_0 path[964531385716605116|+|]_[16982887268023976258|+|] length1863, NZ_CP027534.1|segment_0 path[964531385716605116|+|]_[16982887268023976258|+|] length1863, NZ_CP128947.1|segment_0 path[964531385716605116|+|]_[16982887268023976258|+|] length1863: ISL3, mostly E.coli, same hits
    - NZ_LR890693.1|segment_1 path[964531385716605116|-|]_[16982887268023976258|-|], NZ_CP010876.1|segment_0 path[964531385716605116|-|]_[16982887268023976258|-|] length1893, (order was inverted here) length1893, this one got hits now, all E.coli (why hits even though sequences don't fit)
- NZ_CP124424.1|segment_0 path[964531385716605116|+|] length1324, exist as partly insertion (only first block), less hits, E.coli and Lelliottia amnigena, ISL3, found twice


- NZ_CP021935.1, NZ_CP059130.1 (segment 0), NZ_CP107114.1 and NZ_CP107172.1 (identical), NZ_CP107114.1, NZ_CP107172, NZ_CP021179.1, NZ_CP018979.1 and NZ_CP124429.1 (inverted): 517157241956952768 IS3, 100% hits in E.coli and in other bacteria, mostly E. coli, all get the same hits, they might differ in a few bases or some bases at ends, inserted at different positions in genome, one has 3 base overhang, apart from that they are completely the same, different amount of hits in genome (between 2 and 7)

- NZ_AP026788.1 and NZ_AP026794.1: path[16556873922195442803|+|] IS1380, E. coli, Klebsiella pneumoniae, Pseudomonas aeruginosa,Salmonella enterica all frequent, many more, twice in genome and both have same hits

- NZ_CP019008.1|segment_0 path[185719230501067033|+|] length4204, NZ_CP133923.1|segment_0 path[185719230501067033|+|] length4204, NZ_CP124467.1|segment_0 path[185719230501067033|+|] length4204, NZ_CP134384.1|segment_0 path[185719230501067033|+|] length4203 (one nucleotide shorter): 200 hits that are all E. coli. all sequences identical, only once in genome


Difference between consensus 1 and consensus 2 in one big prophage according to annotations

Be careful to not confuse context for next block!

In [87]:
pd.read_csv("../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus1/NZ_AP022044.1_segment_0.hits_info.tsv", sep="\t")[["query", "qlen", 'hits', 'sgenome', 'qcovGnm', 'cls', 'hsp',
       'qcovHSP', 'alenHSP', 'pident', 'gaps', 'evalue', 'bitscore', 'organism', 'strain',
       'serovar', 'sequence_type']]

query  qlen  hits         sgenome  qcovGnm  cls  hsp  \
0   NZ_AP022044.1|segment_0   767    25    SAMEA4062129    100.0    1    1   
1   NZ_AP022044.1|segment_0   767    25    SAMN14049574    100.0    1    1   
2   NZ_AP022044.1|segment_0   767    25    SAMN07508953    100.0    1    1   
3   NZ_AP022044.1|segment_0   767    25  SAMEA104040230    100.0    1    1   
4   NZ_AP022044.1|segment_0   767    25  SAMEA111271919    100.0    1    1   
5   NZ_AP022044.1|segment_0   767    25    SAMN08382692    100.0    1    1   
6   NZ_AP022044.1|segment_0   767    25  SAMEA111273518    100.0    1    1   
7   NZ_AP022044.1|segment_0   767    25    SAMN18977267    100.0    1    1   
8   NZ_AP022044.1|segment_0   767    25  SAMEA111273646    100.0    1    1   
9   NZ_AP022044.1|segment_0   767    25  SAMEA111273548    100.0    1    1   
10  NZ_AP022044.1|segment_0   767    25  SAMEA111273647    100.0    1    1   
11  NZ_AP022044.1|segment_0   767    25  SAMEA111273667    100.0    1    1   
12  NZ_AP022044.1|segment_0   767    25    SAMN01922332    100.0    1    1   
13  NZ_AP022044.1|segment_0   767    25    SAMN15904810    100.0    1    1   
14  NZ_AP022044.1|segment_0   767    25    SAMN26379194    100.0    1    1   
15  NZ_AP022044.1|segment_0   767    25  SAMEA111274498    100.0    1    1   
16  NZ_AP022044.1|segment_0   767    25  SAMEA111274494    100.0    1    1   
17  NZ_AP022044.1|segment_0   767    25    SAMD00500056    100.0    1    1   
18  NZ_AP022044.1|segment_0   767    25    SAMD00497908    100.0    1    1   
19  NZ_AP022044.1|segment_0   767    25    SAMN15904789    100.0    1    1   
20  NZ_AP022044.1|segment_0   767    25    SAMEA6155701    100.0    1    1   
21  NZ_AP022044.1|segment_0   767    25    SAMEA7230815    100.0    1    1   
22  NZ_AP022044.1|segment_0   767    25    SAMN30575479    100.0    1    1   
23  NZ_AP022044.1|segment_0   767    25    SAMN26796500    100.0    1    1   
24  NZ_AP022044.1|segment_0   767    25    SAMN41034631    100.0    1    1   

    qcovHSP  alenHSP  pident  gaps  evalue  bitscore  \
0     100.0      768  99.870     1     0.0      1389   
1     100.0      768  99.870     1     0.0      1389   
2     100.0      768  99.870     1     0.0      1389   
3     100.0      768  99.870     1     0.0      1389   
4     100.0      768  99.870     1     0.0      1389   
5     100.0      768  99.870     1     0.0      1389   
6     100.0      768  99.870     1     0.0      1389   
7     100.0      768  99.870     1     0.0      1389   
8     100.0      768  99.870     1     0.0      1389   
9     100.0      768  99.870     1     0.0      1389   
10    100.0      768  99.870     1     0.0      1389   
11    100.0      768  99.870     1     0.0      1389   
12    100.0      768  99.870     1     0.0      1389   
13    100.0      768  99.870     1     0.0      1389   
14    100.0      768  99.870     1     0.0      1389   
15    100.0      768  99.870     1     0.0      1389   
16    100.0      768  99.870     1     0.0      1389   
17    100.0      768  99.870     1     0.0      1389   
18    100.0      768  99.870     1     0.0      1389   
19    100.0      768  99.870     1     0.0      1389   
20    100.0      768  99.870     1     0.0      1389   
21    100.0      768  99.740     1     0.0      1386   
22    100.0      768  98.698     1     0.0      1350   
23    100.0      768  98.307     1     0.0      1335   
24    100.0      768  98.177     1     0.0      1332   

                      organism                          strain  serovar  \
0             Escherichia coli                             NaN      NaN   
1             Escherichia coli                E. coli QEC12-15      NaN   
2             Escherichia coli                          29-KPC      NaN   
3             Escherichia coli                             NaN      NaN   
4             Escherichia coli                             NaN      NaN   
5             Escherichia coli                      GASRECE110      NaN   
6          

In [ ]:
# for deletions, retrieve block consensus sequence to search in atb database
# TODO: maybe find a better way that takes all isolates beloning to consensus and containing the block and recalculate consensus sequence (talk to Marco)
# TODO: do I also search deletions in the database?

'GGTAATGACTCCAACTTACTGATAGTGTTTTATGTTCAGATAATGCCCGATGACTTTGTCATGCAGCTCCACCGATTTTGAGAACGACAGTGACTTCCGTCCCAGCCTTGCCAGATGTTGTCTCAGATTCAGGTTATGTCGCTCAATGCGCTGAGTGTAACGCTTGCTGATTACGTGCAGCTTTCCCTTCAGGCGGGATTCATACAGCGGCCAGCCATCCGTCATCCATACCACGACCTCAAAGGCCGACAGCAGGCTCAGAAGACGCTCCAGTGTGGCCAGAGTGCGTTCACCGAAGACGTGCGCCACAACCGTCCTCCGTATCCTGTCATACGCGTAAAACAGCCAGCGCTGACGTGATTTAGCACCGACGTAGCCCCACTGTTCGTCCATTTCAGCGCAGACAATCACATCACTGCCCGGCTGTATGCGCGAGGTTACCGACTGCGGCCTGAGTTTTTTAAGTGACGTAAAATCGTGTTGAGGCCAACGCCCATAATGCGGGCGGTTGCCCGGCATCCAACACCATTCATGGCCATATCAATGATTTTCTGGTGTGTACCGGGTTGAGAAGCGGTGTAAGTGAACTGCAGTTGCCATGTTTTACGGCAGTGAGAGCAGAGATAGCGCTGATGTCCGGCAGTACTTTTACCGTTACGCACCACGCCTTCAGTAGCTGAGCAGGAGGGACAACTGATGGAGATGGAAGCCACGGGAGCACCTCAAAAACACCATCATACACTAAATCAGTAAGTTGGCA'

## Search insertion sequences in their own genome

In [19]:
# make sure that all genome fasta files exist
#convert_gbk_fasta(gbk_folder = "../data/gbk", fasta_folder = "../results/genome_fastas")

In [20]:
insertions_df = find_insertion_hits_own_genome(genome_root = "../results/genome_fastas", insertions_seq_dir="../results/atb_lookup")

[M::mm_idx_gen::0.075*1.21] collected minimizers
[M::mm_idx_gen::0.086*1.40] sorted minimizers
[M::main::0.086*1.40] loaded/built the index for 1 target sequence(s)
[M::mm_mapopt_update::0.091*1.37] mid_occ = 50
[M::mm_idx_stat] kmer size: 19; skip: 19; is_hpc: 0; #seq: 1
[M::mm_idx_stat::0.096*1.35] distinct minimizers: 504461 (98.70% are singletons); average occurrences: 1.024; average spacing: 9.992; total length: 5163545
[M::worker_pipeline::0.097*1.35] mapped 1 sequences
[M::main] Version: 2.30-r1287
[M::main] CMD: minimap2 -x asm5 -N 50 -p 0.9 -k 19 --eqx ../results/genome_fastas/NZ_CP124455.1.fasta ../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus1/NZ_CP124455.1_segment_0.fasta
[M::main] Real time: 0.103 sec; CPU: 0.137 sec; Peak RSS: 0.043 GB
[M::mm_idx_gen::0.064*1.04] collected minimizers
[M::mm_idx_gen::0.073*1.25] sorted minimizers
[M::main::0.073*1.25] loaded/built the index for 1 target sequence(s)
[M::mm_mapopt_update::0.078*1.24] mid_occ = 50
[M::mm_idx_stat] k

In [21]:
insertions_df.sort_values(['consensus','insertion_path'])

junction_name   consensus    genome_name  \
27  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP026794.1   
28  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP026788.1   
9   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124467.1   
29  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP134384.1   
36  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP133923.1   
41  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP019008.1   
12  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP021935.1   
13  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107114.1   
14  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107172.1   
18  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP059130.1   
19  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP021179.1   
10  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124429.1   
16  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP018979.1   
15  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP051659.1   
39  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP051661.1   
17  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107122.1   
30  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP051661.1   
34  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP022044.1   
32  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP030337.1   
26  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP128907.1   
24  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP022326.1   
42  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP022326.1   
45  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP051663.1   
33  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124410.1   
35  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP019012.1   
5   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_LR890693.1   
21  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_OX030701.1   
48  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP022044.1   
0   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124455.1   
1   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107122.1   
23  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107162.1   
44  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107151.1   
7   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP023826.1   
20  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP018970.1   
8   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP021454.1   
47  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP103562.1   
31  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124424.1   
2   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP128947.1   
3   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP059130.1   
4   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124460.1   
22  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP076693.1   
25  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP036245.1   
37  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP026794.1   
38  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP027534.1   
40  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP026788.1   
6   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP010876.1   
11  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_LR890693.1   
43  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124374.1   
46  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP019015.1   
50  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus2  NZ_AP022171.1   
49  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus2  NZ_CP098219.1   

                                     insertion_path  insertion_length  \
27                          16556873922195442803|+|              1661   
28                          16556873922195442803|+|              1661   
9                             185719230501067033|+|              4204   
29                            185719230501067033|+|              4203   
36                            185719230501067033|+|              4204   
41                            185719230501067033|+|              4204   
12                            517157241956952768|+|              1248   
13                            517157241956952768|+|              1248   
14                            517157241956952768|+|              1248   
18                            517157241956952768|+|              1253   
19                            51715724195

In [22]:
# TODO: search in plasmids as well
# convert_gbk_fasta_plasmids(gbk_folder = "../data/plasmids", fasta_folder = "../results/plasmid_fastas")

In [23]:
plasmid_insertions_df = find_insertion_hits_in_plasmids(plasmid_fasta_root="../results/plasmid_fastas", insertions_seq_dir="../results/atb_lookup")

[M::mm_idx_gen::0.001*8.29] collected minimizers
[M::mm_idx_gen::0.001*5.99] sorted minimizers
[M::main::0.001*5.97] loaded/built the index for 1 target sequence(s)
[M::mm_mapopt_update::0.001*5.81] mid_occ = 50
[M::mm_idx_stat] kmer size: 19; skip: 19; is_hpc: 0; #seq: 1
[M::mm_idx_stat::0.001*5.65] distinct minimizers: 415 (100.00% are singletons); average occurrences: 1.000; average spacing: 9.848; total length: 4087
[M::worker_pipeline::0.001*5.07] mapped 1 sequences
[M::main] Version: 2.30-r1287
[M::main] CMD: minimap2 -x asm5 -N 50 -p 0.9 -k 19 --eqx ../results/plasmid_fastas/NZ_CP124455.1/NZ_CP124458.1.fasta ../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/consensus1/NZ_CP124455.1_segment_0.fasta
[M::main] Real time: 0.002 sec; CPU: 0.007 sec; Peak RSS: 0.002 GB
[M::mm_idx_gen::0.000*10.67] collected minimizers
[M::mm_idx_gen::0.001*7.27] sorted minimizers
[M::main::0.001*7.24] loaded/built the index for 1 target sequence(s)
[M::mm_mapopt_update::0.001*6.96] mid_occ = 50
[M::mm_

In [24]:
plasmid_insertions_df

junction_name   consensus   isolate_name   plasmid_name  \
0      atb_lookup  consensus1  NZ_CP124455.1  NZ_CP124458.1   
1      atb_lookup  consensus1  NZ_CP124455.1  NZ_CP124459.1   
2      atb_lookup  consensus1  NZ_CP124455.1  NZ_CP124456.1   
3      atb_lookup  consensus1  NZ_CP124455.1  NZ_CP124457.1   
4      atb_lookup  consensus1  NZ_CP107122.1  NZ_CP107124.1   
..            ...         ...            ...            ...   
170    atb_lookup  consensus1  NZ_AP022044.1  NZ_AP022046.1   
171    atb_lookup  consensus2  NZ_CP098219.1  NZ_CP098222.1   
172    atb_lookup  consensus2  NZ_CP098219.1  NZ_CP098221.1   
173    atb_lookup  consensus2  NZ_CP098219.1  NZ_CP098220.1   
174    atb_lookup  consensus2  NZ_AP022171.1  NZ_AP022172.1   

                                 insertion_path  insertion_length    segment  \
0    6932368721519041699|-|15558722379768673797               768  segment_0   
1    6932368721519041699|-|15558722379768673797               768  segment_0   
2    6932368721519041699|-|15558722379768673797               768  segment_0   
3    6932368721519041699|-|15558722379768673797               768  segment_0   
4    6932368721519041699|-|15558722379768673797               777  segment_0   
..                                          ...               ...        ...   
170  6932368721519041699|-|14506298656834147691               767  segment_1   
171                      2278011600046483881|+|             37622  segment_0   
172                      2278011600046483881|+|             37622  segment_0   
173                      2278011600046483881|+|             37622  segment_0   
174                     18228278273347143766|+|               820  segment_0   

     hits_in_plasmid  
0                  0  
1                  0  
2                  1  
3                  0  
4                  0  
..               ...  
170                0  
171                0  
172                0  
173                0  
174                0  

[175 rows x 8 columns]

In [25]:
plasmid_summary_df = plasmid_insertions_df.groupby(["consensus", "isolate_name"])["hits_in_plasmid"].sum().reset_index()
insertions_df = insertions_df.merge(
    plasmid_summary_df,
    left_on=["consensus", "genome_name"],
    right_on=["consensus", "isolate_name"],
    how="left"
).drop(columns=['isolate_name'])

insertions_df.sort_values(['consensus','insertion_path'])

junction_name   consensus    genome_name  \
27  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP026794.1   
28  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP026788.1   
9   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124467.1   
29  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP134384.1   
36  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP133923.1   
41  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP019008.1   
12  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP021935.1   
13  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107114.1   
14  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107172.1   
18  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP059130.1   
19  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP021179.1   
10  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124429.1   
16  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP018979.1   
15  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP051659.1   
39  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP051661.1   
17  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107122.1   
30  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP051661.1   
34  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP022044.1   
32  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP030337.1   
26  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP128907.1   
24  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP022326.1   
42  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP022326.1   
45  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP051663.1   
33  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124410.1   
35  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP019012.1   
5   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_LR890693.1   
21  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_OX030701.1   
48  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP022044.1   
0   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124455.1   
1   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107122.1   
23  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107162.1   
44  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP107151.1   
7   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP023826.1   
20  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP018970.1   
8   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP021454.1   
47  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP103562.1   
31  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124424.1   
2   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP128947.1   
3   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP059130.1   
4   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124460.1   
22  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP076693.1   
25  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP036245.1   
37  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP026794.1   
38  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP027534.1   
40  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_AP026788.1   
6   CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP010876.1   
11  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_LR890693.1   
43  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP124374.1   
46  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus1  NZ_CP019015.1   
50  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus2  NZ_AP022171.1   
49  CIRMBUYJFK_f__CWCCKOQCWZ_r  consensus2  NZ_CP098219.1   

                                     insertion_path  insertion_length  \
27                          16556873922195442803|+|              1661   
28                          16556873922195442803|+|              1661   
9                             185719230501067033|+|              4204   
29                            185719230501067033|+|              4203   
36                            185719230501067033|+|              4204   
41                            185719230501067033|+|              4204   
12                            517157241956952768|+|              1248   
13                            517157241956952768|+|              1248   
14                            517157241956952768|+|              1248   
18                            517157241956952768|+|              1253   
19                            51715724195

In [27]:
insertions_df.to_csv("../results/atb_lookup/CIRMBUYJFK_f__CWCCKOQCWZ_r/insertions_df.tsv", sep="\t", index=False)